### ISSUE TITLE
Plugin hook to customize __repr__ of an Instance

### ISSUE THREAD

**Issue by mkurnikov** [python/mypy#6501]
For example, in Django, there's this `QuerySet` object with a lot of methods, and to properly support those methods we need two (and maybe more in the future) generic parameters. 

With `get_type_analyze_hook` I can convert `QuerySet[MyModel]` into `QuerySet[MyModel, MyModel]`, so in the code user don't need to think about number of generics. But in error messages, all those generics values are present, though they don't have any semantics, it's just implementation detail. 

Or, for `Field`, I need to have two generic params, one for type of `__set__`, one for `__get__`, so they appear in error message as (`CharField` for example) 
```
django.db.models.fields.CharField[Union[builtins.str, builtins.int, django.db.models.expressions.Combinable], builtins.str]
```
None of this info makes any sense to users, so I want it to just be `django.db.models.fields.CharField`.

This is monkeypatching version that works for one of cases, should be simple enough to add hook for that
```
def make_queryset_repr_return_only_one_type():
    from mypy.types import TypeStrVisitor

    old_visit_instance = TypeStrVisitor.visit_instance

    def patched_visit_instance(self, t: Instance) -> str:
        if t.type.has_base(helpers.QUERYSET_CLASS_FULLNAME):
            return old_visit_instance(self, helpers.reparametrize_instance(t, [t.args[0]]))
        else:
            return old_visit_instance(self, t)

    TypeStrVisitor.visit_instance = patched_visit_instance
```

**Comment by ilevkivskyi** [comment]
Yes, this looks like a reasonable feature. The main problem with implementing will be likely avoiding import cycles.
